Sources/protoc-gen-swift/ReservedWords.swift

```diff
@@ -47,47 +46,45 @@ private let reservedTypeNames: Set<String> = {
     return names
 }()
 
-func sanitizeMessageTypeName(_ s: String) -> String {
+private func sanitizeTypeName(_ s: String, disambiguator: String) -> String {
     if reservedTypeNames.contains(s) {
-        return s + "Message"
+        return s + disambiguator
     } else if isAllUnderscore(s) {
-        return s + "Message"
+        return s + disambiguator
+    } else if s.hasSuffix(disambiguator) {
+        let e = s.index(s.endIndex, offsetBy: -disambiguator.characters.count)
+        let truncated = s.substring(to: e)
+        return sanitizeTypeName(truncated, disambiguator: disambiguator) + disambiguator
     } else {
         return s
     }
 }
 
+func sanitizeMessageTypeName(_ s: String) -> String {
+    return sanitizeTypeName(s, disambiguator: "Message")
+}
 
 func sanitizeEnumTypeName(_ s: String) -> String {
-    if reservedTypeNames.contains(s) {
-        return s + "Enum"
-    } else if isAllUnderscore(s) {
-        return s + "Enum"
-    } else {
-        return s
-    }
+    return sanitizeTypeName(s, disambiguator: "Enum")
 }
 
 func sanitizeOneofTypeName(_ s: String) -> String {
-    if reservedTypeNames.contains(s) {
-        return s + "Oneof"
-    } else if isAllUnderscore(s) {
-        return s + "Oneof"
-    } else {
-        return s
-    }
+    return sanitizeTypeName(s, disambiguator: "Oneof")
 }
 
 private let reservedFieldNames: Set<String> =  {
     () -> Set<String> in
     var names: Set<String> = [
-            "debugDescription",
-            "description",
-            "dynamicType",
-            "hashValue",
-            "isInitialized",
-            "jsonFieldNames",
-            "unknownFields",
+        "debugDescription",
+        "decoder",
+        "description",
+        "dynamicType",
+        "fieldNumber",
+        "hashValue",
+        "isInitialized",
+        "other",
```

Comment:
Should we try to be smarter here and have a list of words that are just locals like this, and prepend `self.` to their references in those functions? That way we don't waste names that otherwise don't have to be lost otherwise.